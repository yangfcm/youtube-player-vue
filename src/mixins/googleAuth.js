import { mapActions } from "vuex";
const googleAuth = {
  created() {
    if (!window.gapi) return;
    window.gapi.load("client:auth2", () => {
      window.gapi.client
        .init({
          clientId: process.env.VUE_APP_CLIENT_ID,
          scope:
            "email https://www.googleapis.com/auth/userinfo.profile https://www.googleapis.com/auth/youtube https://www.googleapis.com/auth/youtube.force-ssl",
        })
        .then(() => {
          this.auth = window.gapi.auth2.getAuthInstance();
          this.onAuthChange(this.auth.isSignedIn.get());
          this.auth.isSignedIn.listen(this.onAuthChange);
        });
    });
  },

  methods: {
    ...mapActions({
      signIn: "signIn",
      signOut: "signOut",
    }),
    onAuthChange(isSignedIn) {
      if (isSignedIn) {
        this.signIn(this.auth.currentUser.get()); // signIn action
      } else {
        this.signOut(); // signOut action
      }
    },
    handleGoogleSignout() {
      this.auth.signOut(); // signOut provided by Google Auth API
    },
    handleGoogleSignin() {
      this.auth.signIn(); // signIn provided by Google Auth API
    },
  },
};

export default googleAuth;
