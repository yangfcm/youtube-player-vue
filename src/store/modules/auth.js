import axios from "../../settings";
import { SIGN_IN, SIGN_OUT, CATCH_AUTH_ERROR } from "../types";

const state = {
  user: null,
};

const getters = {
  signedIn: (state) => {
    const { user } = state;
    if (user === null) {
      return null;
    }
    if (user) {
      if (Object.keys(user).length === 0) {
        return false;
      } else {
        return true;
      }
    }
  },
  user: (state) => state.user,
};

const mutations = {
  SIGN_IN: (state, payload) => {
    state.user = payload;
  },
  SIGN_OUT: (state) => {
    state.user = {};
  },
  CATCH_AUTH_ERROR: (state) => {
    state.user = {};
  },
};

const actions = {
  signIn: async (context, googleUser) => {
    try {
      const accessToken = `${googleUser.wc.token_type} ${googleUser.wc.access_token}`;
      localStorage.setItem("access_token", accessToken);
      const response = await axios.get(
        "https://www.googleapis.com/oauth2/v1/userinfo?alt=json",
        {
          headers: {
            Authorization: accessToken,
          },
        }
      );
      context.commit(SIGN_IN, response.data);
    } catch (err) {
      context.commit(CATCH_AUTH_ERROR);
    }
  },
  signOut: (context) => {
    localStorage.removeItem("access_token");
    context.commit(SIGN_OUT);
  },
};

export default {
  state,
  getters,
  mutations,
  actions,
};
