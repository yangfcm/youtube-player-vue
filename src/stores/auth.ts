import { ref, computed } from 'vue'
import { defineStore } from 'pinia'
import { type AxiosResponse } from 'axios'
import { AsyncStatus } from '@/settings/types'
import { gAuthAxios } from '@/settings/api'
import type { PlayListsRespone, SubscriptionsResponse } from './types'
import { fetchSubscriptionsAPI, fetchPlayListsAPI } from './api'

type UserInfoResponse = {
  email: string
  email_verified: boolean
  family_name: string
  given_name: string
  locale: string
  name: string
  picture: string
  sub: string
}

type Auth = {
  user: UserInfoResponse | null
  status: AsyncStatus
  subscriptions: {
    status: AsyncStatus
    error: string
    data?: SubscriptionsResponse
    subscriptionIds: Record<string, string>
  }
  playlists: {
    status: AsyncStatus
    error: string
    data?: PlayListsRespone
  }
}

export const useAuthStore = defineStore('auth', () => {
  const auth = ref<Auth>({
    user: null,
    status: AsyncStatus.IDLE,
    subscriptions: {
      status: AsyncStatus.IDLE,
      error: '',
      subscriptionIds: {},
    },
    playlists: {
      status: AsyncStatus.IDLE,
      error: '',
    },
  })

  // function signin(authData: Auth) {
  //   auth.value = authData
  // }

  function signout() {
    auth.value.user = null
    auth.value.status = AsyncStatus.IDLE
    auth.value.subscriptions = {
      status: AsyncStatus.IDLE,
      error: '',
      subscriptionIds: {},
    }
    auth.value.playlists = {
      status: AsyncStatus.IDLE,
      error: '',
    }
  }

  const status = computed(() => {
    return auth.value.status
  })

  const user = computed(() => {
    return auth.value.user
  })

  const isSignedIn = computed(() => {
    if (!auth.value.user) return false
    return true
  })

  const subscriptions = computed(() => {
    return auth.value.subscriptions
  })

  const playlists = computed(() => {
    return auth.value.playlists
  })

  const fetchUserByToken = async (token: string) => {
    try {
      auth.value.status = AsyncStatus.LOADING
      const response: AxiosResponse<UserInfoResponse> = await gAuthAxios.get('/', {
        params: {
          access_token: token,
        },
      })
      auth.value.user = response.data
      auth.value.status = AsyncStatus.SUCCESS
    } catch {
      signout()
    }
  }

  const fetchSubscriptions = async (pageToken?: string) => {
    try {
      auth.value.subscriptions.status = AsyncStatus.LOADING
      auth.value.subscriptions.error = ''
      const options: Record<string, string> = {}
      if (pageToken) options.pageToken = pageToken
      const response = await fetchSubscriptionsAPI(options)
      const currentItems = auth.value.subscriptions.data?.items || []
      auth.value.subscriptions.status = AsyncStatus.SUCCESS
      auth.value.subscriptions.data = {
        ...response.data,
        items: [...currentItems, ...response.data.items],
      }
      response.data.items.forEach(
        (item) =>
          (auth.value.subscriptions.subscriptionIds[item.snippet.resourceId.channelId] = item.id),
      )
    } catch (err: any) {
      auth.value.subscriptions.status = AsyncStatus.FAIL
      auth.value.subscriptions.error = err.message
    }
  }

  const fetchPlayLists = async (pageToken?: string) => {
    try {
      auth.value.playlists.status = AsyncStatus.LOADING
      auth.value.playlists.error = ''
      const options: Record<string, string> = {}
      if (pageToken) options.pageToken = pageToken
      const response = await fetchPlayListsAPI(options)
      const currentItems = auth.value.playlists.data?.items || []
      auth.value.playlists.status = AsyncStatus.SUCCESS
      auth.value.playlists.data = {
        ...response.data,
        items: [...currentItems, ...response.data.items],
      }
    } catch (err: any) {
      auth.value.playlists.status = AsyncStatus.FAIL
      auth.value.playlists.error = err.message
    }
  }

  return {
    status,
    user,
    isSignedIn,
    subscriptions,
    playlists,
    // signin,
    signout,
    fetchUserByToken,
    fetchSubscriptions,
    fetchPlayLists,
    authState: auth,
  }
})
