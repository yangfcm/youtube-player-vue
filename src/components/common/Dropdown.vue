<template>
  <div class="ui vertical menu">
    <router-link class="item" to="/">Recommend</router-link>
    <router-link class="item" to="/subscription">Subscriptions</router-link>
    <router-link class="item" to="/playlist">Play list</router-link>
    <a class="item" @click="handleGoogleSignout">Sign out</a>
  </div>
</template>

<script>
import googleAuth from "../../mixins/googleAuth";
export default {
  mixins: [googleAuth]
};
</script>

<style scoped>
</style>