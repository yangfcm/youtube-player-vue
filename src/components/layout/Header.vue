<template>
  <div class="ui menu stackable">
    <router-link to="/" class="header item ui">
      <i class="video icon red"></i>
      FanTube
    </router-link>
    <div class="right menu">
      <div class="item">
        <div class="ui icon input" style="min-width: 10px;">
          <input type="text" />
          <i class="search icon link"></i>
        </div>
        <div v-if="signedIn === true" class="app-dropdown-container">
          <img
            :src="user.picture"
            :alt="user.name"
            class="ui tiny image circular"
            style="cursor: pointer; max-height: 60px; width: auto;"
            @click.stop="toggleShowDropdown"
          />
          <app-dropdown class="app-dropdown-menu" v-if="showDropdown"></app-dropdown>
        </div>
        <div v-if="signedIn === false">
          <button
            class="ui youtube button"
            style="margin-left: 1rem; width: 110px;"
            @click="handleSignin"
          >
            <i class="youtube icon"></i>
            Sign In
          </button>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import { mapState, mapGetters } from "vuex";
import googleAuth from "../../mixins/googleAuth";
import Dropdown from "../common/Dropdown";

export default {
  data: function() {
    return {
      showDropdown: false
    };
  },
  components: {
    appDropdown: Dropdown
  },
  computed: {
    ...mapGetters(["user", "signedIn"])
  },
  methods: {
    toggleShowDropdown() {
      this.showDropdown = !this.showDropdown;
    }
  },
  mounted() {
    window.addEventListener("click", () => {
      this.showDropdown = false;
    });
  },
  mixins: [googleAuth]
};
</script>

<style scoped>
.app-dropdown-container {
  position: relative;
  margin-left: 1rem;
}
.app-dropdown-menu {
  position: absolute;
  right: 0;
  top: 105%;
  z-index: 10;
}
</style>
