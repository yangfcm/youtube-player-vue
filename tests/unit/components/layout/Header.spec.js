import { mount, RouterLinkStub } from "@vue/test-utils";
import Header from "@/components/layout/Header.vue";
import Dropdown from "@/components/common/Dropdown.vue";
import { userResponse as user } from "../../store/fixtures/auth";
import googleAuth from "@/mixins/googleAuth";

jest.mock("@/mixins/googleAuth");

describe("Test Header component", () => {
  let headerWrapper, authedHeaderWrapper;

  beforeEach(() => {
    headerWrapper = mount(Header, {
      mocks: {
        $store: {
          getters: {
            user: {},
            signedIn: false,
          },
        },
      },
      stubs: {
        RouterLink: RouterLinkStub,
      },
    });

    authedHeaderWrapper = mount(Header, {
      mocks: {
        $store: {
          getters: {
            user,
            signedIn: true,
          },
        },
      },
      stubs: {
        RouterLink: RouterLinkStub,
      },
    });
  });
  afterEach(() => {
    headerWrapper.destroy();
    authedHeaderWrapper.destroy();
  });

  it("should render Header component correctly", () => {
    expect(headerWrapper).toMatchSnapshot();
    expect(authedHeaderWrapper).toMatchSnapshot();
  });

  it("Header link should navigate to homepage", () => {
    expect(headerWrapper.findComponent(RouterLinkStub).props().to).toBe("/");
    expect(authedHeaderWrapper.findComponent(RouterLinkStub).props().to).toBe(
      "/"
    );
  });

  it("Header should have sign in button, but not have avatar image", () => {
    expect(headerWrapper.find("button").exists()).toBe(true);
    expect(headerWrapper.find("button").text()).toContain("Sign In");
    expect(headerWrapper.find("img").exists()).toBe(false);
  });

  it("Authed Header should have avatar image, but no sign in button", () => {
    expect(authedHeaderWrapper.find("button").exists()).toBe(false);
    expect(authedHeaderWrapper.find("img").exists()).toBe(true);
  });

  it("Click sign in button can call handleSignin", () => {
    headerWrapper.find("button").trigger("click");
    // const mockTest = jest.spyOn(headerWrapper.vm, "handleSignin");
    // expect(mockTest).toHaveBeenCalled();
    // It doesn't work so I use the approach below.
    expect(googleAuth.methods.handleGoogleSignin).toHaveBeenCalled();
  });

  it("Click user avatar can open dropdown and click again can close dropdown", async () => {
    // Initial state
    expect(authedHeaderWrapper.vm.showDropdown).toBe(false);
    expect(authedHeaderWrapper.findComponent(Dropdown).exists()).toBe(false);
    // Click once, open dropdown
    await authedHeaderWrapper.find("img").trigger("click");
    expect(authedHeaderWrapper.vm.showDropdown).toBe(true);
    expect(authedHeaderWrapper.findComponent(Dropdown).exists()).toBe(true);

    // Click again, close dropdown
    await authedHeaderWrapper.find("img").trigger("click");
    expect(authedHeaderWrapper.vm.showDropdown).toBe(false);
    expect(authedHeaderWrapper.findComponent(Dropdown).exists()).toBe(false);
  });

  it("User input can change keyword", () => {
    headerWrapper.find("input").setValue("popular movie");
    expect(headerWrapper.vm.searchKeyword).toBe("popular movie");
  });
});
