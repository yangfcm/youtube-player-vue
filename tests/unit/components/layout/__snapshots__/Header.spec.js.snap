// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Header component should render Header component correctly 1`] = `
<div class="ui menu stackable"><a class="header item ui"><i class="video icon red"></i>
    FanTube
  </a>
  <div class="right menu">
    <div class="item">
      <div class="ui icon input" style="min-width: 10px;"><input type="text"> <i class="search icon link"></i></div>
      <!---->
      <div><button class="ui youtube button" style="margin-left: 1rem; width: 110px;"><i class="youtube icon"></i>
          Sign In
        </button></div>
    </div>
  </div>
</div>
`;

exports[`Test Header component should render Header component correctly 2`] = `
<div class="ui menu stackable"><a class="header item ui"><i class="video icon red"></i>
    FanTube
  </a>
  <div class="right menu">
    <div class="item">
      <div class="ui icon input" style="min-width: 10px;"><input type="text"> <i class="search icon link"></i></div>
      <div class="app-dropdown-container"><img src="https://lh5.googleusercontent.com/-MuN5BmJ3lFM/AAAAAAAAAAI/AAAAAAAAAAA/AMZuucmzrAOkewOEoGUPasQOSYSk3Gl9nA/photo.jpg" alt="sam test" class="ui tiny image circular" style="cursor: pointer; max-height: 60px; width: auto;">
        <!---->
      </div>
      <!---->
    </div>
  </div>
</div>
`;
